### PR TITLE
Masking options

### PR DESCRIPTION
**Pull request by hpparvi**
This PR offers one way to include the mask-handling functionality without requiring a major refactoring of the `_ImageParser` class and all the code that depends on it.  I've changed the `_mask_and_nonfinite_data_handling` method into a static method that takes the cross-dispersion axis and mask treatment as optional arguments (that I've added to the `_get_data_from_image` and `_parse_image` methods as well),  and I've changed all the code using the methods to include this information when necessary. 

I've tried to take a minimalistic approach and modify only what I think is really necessary to make the mask treatment work with the current `_ImageParser` architecture. I'm afraid we will still need to do a major refactoring of image parsing, but this will require a bit of discussion and can be done in a dedicated PR. 

The approach passes all the tests without failures (including codestyle). I also made a minor change in `test_tracing.py` to test that a warning is raised using `pytest.warns` instead of `pytest.raises`, and  merged the recent contributions to the main repo and branch by pllim.

### CODE REVIEW

**Comment by hpparvi** [comment]
I've reverted the PR to a commit that doesn't include the merge with the main repo.

File: specreduce/core.py [hunk 1]
```diff
@@ -28,22 +28,32 @@ class _ImageParser:
...
+        image : `~astropy.nddata.NDData`-like or array-like
             The image to be parsed. If None, defaults to class' own
             image attribute.
-        disp_axis : int, optional
+        disp_axis
             The index of the image's dispersion axis. Should not be
             changed until operations can handle variable image
             orientations. [default: 1]
-        """
+        mask_treatment
```
Comment (cshanahan1) [inline specreduce/core.py:50]: can you add the possible options to the docstring?

Comment (cshanahan1) [inline specreduce/core.py:50]: can you add the possible options and default to the doc string?

File: specreduce/core.py [hunk 2]
```diff
@@ -124,31 +146,33 @@ def _mask_and_nonfinite_data_handling(self, image, mask):
         of masked and nonfinite data - filter, omit, and zero-fill.
         Depending on the step, all or a subset of these three options are valid.
 
+        Parameters
+        ----------
+        image : array-like
+            The input image data array that may contain nonfinite values.
+        mask : array-like or None
```
Comment (cshanahan1) [inline specreduce/core.py]: maybe clarify that this is a boolean mask, since we were talking about expanding this at some point

**Review by cshanahan1 — COMMENTED** [review]
left a few comments, testing it out then i will leave any further comments or an approval. thanks for doing this!

**Review by cshanahan1 — APPROVED** [review]

**Pull request merged**